### PR TITLE
add new seq2seq example: rnn_search

### PR DESCRIPTION
**Pull request by 1024er**
new seq2seq API example: rnn search

### CODE REVIEW

File: PaddleNLP/PaddleTextGEN/rnn_search/util.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+
+def is_sequence(seq):
+    if isinstance(seq, dict):
+        return True
+    return (isinstance(seq, collections.Sequence)
+            and not isinstance(seq, six.string_types))
+
+
+def _sorted(dict_):
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/util.py]: 这个util.py文件里的内容都已经合入paddle了，可以删掉

File: PaddleNLP/PaddleTextGEN/rnn_search/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    --max_grad_norm 5.0 \
+    --vocab_prefix data/en-vi/vocab \
+    --infer_file data/en-vi/tst2013.en \
+    --reload_model attention_models/epoch_10/ \
+    --infer_output_file attention_infer_output/infer_output.txt \
+    --beam_size 10 \
+    --use_gpu True
+```
+
+## 效果
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/README.md]: 加上如何进行评估的说明

File: PaddleNLP/PaddleTextGEN/rnn_search/README.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+## 模型训练
+
+`run.sh`包含训练程序的主函数，要使用默认参数开始训练，只需要简单地执行：
+
+```
+sh run.sh
+```
+
+默认使用带有注意力机制的RNN模型，可以通过修改`--attention' 为False来训练不带注意力机制的RNN模型。
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/README.md]: 这里格式有些问题

File: PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    def __init__(self, num_layers, hidden_size, dropout_prob=0.,
+                 init_scale=0.1):
+        super(AttentionDecoderCell, self).__init__(num_layers, hidden_size,
+                                                   dropout_prob, init_scale)
+        '''
+        self.attention_weight = layers.create_parameter([self.hidden_size * 2, self.hidden_size], dtype="float32", name="attention_weight", \
+                default_initializer=fluid.initializer.UniformInitializer(low=-init_scale, high=init_scale))
+        self.memory_weight = layers.create_parameter([self.hidden_size, self.hidden_size], dtype="float32", name="memory_weight", \
+                default_initializer=fluid.initializer.UniformInitializer(low=-init_scale, high=init_scale))
+        '''
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py]: 是否需要删除

File: PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                                sequence_length=None,
+                                enc_output=self.enc_output,
+                                enc_padding_mask=enc_padding_mask)
+
+            dec_output = output_layer(dec_output)
+
+        elif mode == 'beam_search':
+            start_token = 1
+            end_token = 2
+            max_length = 100
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py]: 这三个不建议硬编码

File: PaddleNLP/PaddleTextGEN/rnn_search/util.py (hunk 1 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/util.py]: 已修复

File: PaddleNLP/PaddleTextGEN/rnn_search/README.md (hunk 2 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/README.md]: 已加上

File: PaddleNLP/PaddleTextGEN/rnn_search/base_model.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                                              inputs=self.tar_emb,
+                                              initial_states=enc_final_state)
+
+            dec_output = output_layer(dec_output)
+
+            return dec_output
+        elif mode == 'beam_search':
+            start_token = 1
+            end_token = 2
+            max_length = 100
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/base_model.py]: 这三个不建议硬编码

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+            total_loss = 0
+            word_count = 0.0
+            batch_times = []
+            for batch_id, batch in enumerate(train_data_iter):
+                batch_start_time = time.time()
+                input_data_feed, word_num = prepare_input(
+                    batch, epoch_id=epoch_id)
+                fetch_outs = exe.run(program=train_program,
+                                     feed=input_data_feed,
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/train.py:198]: 建议使用 compiledprogram （当前这种用法无法支持多卡）和 dataloader，可以参考这里 https://github.com/PaddlePaddle/book/blob/for_paddle1.6/08.machine_translation/seq2seq.py#L287

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+                                     fetch_list=[loss.name],
+                                     use_program_cache=True)
+
+                cost_train = np.array(fetch_outs[0])
+                # print(cost_train)
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/train.py:203]: 删掉注释掉的可删掉的代码

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+            end_time = time.time()
+            epoch_time = end_time - start_time
+            ce_time.append(epoch_time)
+            print(
+                "\nTrain epoch:[%d]; Epoch Time: %.5f; avg_time: %.5f s/step\n"
+                % (epoch_id, epoch_time, sum(batch_times) / len(batch_times)))
+
+            if not args.profile:
+                dir_name = args.model_path + "/epoch_" + str(epoch_id)
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/train.py]: 这个路径中的分隔符应该会有Windows不兼容的问题，要修改下

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+    optimizer.minimize(loss)
+
+    place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
+    exe = Executor(place)
+    exe.run(framework.default_startup_program())
+
+    device_count = len(fluid.cuda_places()) if args.use_gpu else len(
+        fluid.cpu_places())
+    if device_count > 1:
+        raise Exception("Training using multi-GPUs is not supported now.")
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/train.py]: 也请添加下多卡支持，参考这里 https://github.com/PaddlePaddle/book/blob/for_paddle1.6/08.machine_translation/seq2seq.py#L280

File: PaddleNLP/PaddleTextGEN/rnn_search/args.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    parser.add_argument(
+        "--enable_ce",
+        action='store_true',
+        help="The flag indicating whether to run the task "
+        "for continuous evaluation.")
+
+    parser.add_argument(
+        "--parallel",
+        action='store_true',
+        help="Whether execute with the data_parallel mode.")
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/args.py]: 都使用compiledprogram后，parallel这个选项可以去掉

File: PaddleNLP/PaddleTextGEN/rnn_search/args.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    parser.add_argument(
+        "--max_grad_norm",
+        type=float,
+        default=5.0,
+        help="max grad norm for global norm clip")
+
+    parser.add_argument(
+        "--model_path",
+        type=str,
+        default='./model',
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/rnn_search/args.py]: 注意可能存在的windows兼容性问题

File: PaddleNLP/PaddleTextGEN/vae_text/args.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    parser.add_argument(
+        "--enable_ce",
+        action='store_true',
+        help="The flag indicating whether to run the task "
+        "for continuous evaluation.")
+
+    parser.add_argument(
+        "--parallel",
+        action='store_true',
+        help="Whether execute with the data_parallel mode.")
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/vae_text/args.py]: 使用CompiledProgram后，这个选项可以去掉

File: PaddleNLP/PaddleTextGEN/vae_text/infer.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    beam_size = args.beam_size
+    trans_res = model.build_graph(mode='sampling', beam_size=beam_size)
+    # clone from default main program and use it as the validation program
+    main_program = fluid.default_main_program()
+
+    place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
+    exe = Executor(place)
+    exe.run(framework.default_startup_program())
+
+    dir_name = './' + args.reload_model
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/vae_text/infer.py]: 注意Windows兼容性问题

File: PaddleNLP/PaddleTextGEN/vae_text/infer.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    print("dir name", dir_name)
+    fluid.io.load_params(exe, dir_name)
+    vocab, tar_id2vocab = get_vocab(args.dataset_prefix)
+    infer_output = np.ones((batch_size, 1), dtype='int64') * BOS_ID
+
+    fetch_outs = exe.run(feed={'tar': infer_output},
+                         fetch_list=[trans_res.name],
```
Comment (guoshengCS) [inline PaddleNLP/PaddleTextGEN/vae_text/infer.py:96]: 使用CompiledProgram，可以的话也请使用dataloader ，参考 https://github.com/PaddlePaddle/book/blob/for_paddle1.6/08.machine_translation/seq2seq.py#L287

File: PaddleNLP/PaddleTextGEN/rnn_search/reader.py [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+UNK_ID = 0
+
+
+def _read_words(filename):
+    data = []
+    with io.open(filename, "r", encoding='utf-8') as f:
+        if Py3:
+            return f.read().replace("\n", "<eos>").split()
+        else:
+            return f.read().decode("utf-8").replace("\n", "<eos>").split()
```
Comment (xyzhou-puck) [inline PaddleNLP/PaddleTextGEN/rnn_search/reader.py]: u“\n” and u"<eos>"

File: PaddleNLP/PaddleTextGEN/rnn_search/reader.py [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+            if not source:
+                mask[i] = len(ele) - 1
+            else:
+                mask[i] = len(ele)
+
+        return ids, mask
+
+    b_src = []
+
+    cache_num = 20
```
Comment (xyzhou-puck) [inline PaddleNLP/PaddleTextGEN/rnn_search/reader.py]: 尽量不要hard code

**Review by xyzhou-puck — APPROVED** [review]
LGTM.

File: PaddleNLP/PaddleTextGEN/rnn_search/args.py (hunk 12 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/args.py]: 已经修改

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py (hunk 7 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/train.py:198]: RNN Search已经支持多卡；VAE Text 需要较大改动，放到后续版本支持多卡

File: PaddleNLP/PaddleTextGEN/vae_text/args.py (hunk 13 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/vae_text/args.py]: 已经fix

File: PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py (hunk 5 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/attention_model.py]: 写入到了__init__()中了，设置为默认值
class AttentionModel(BaseModel):
    def __init__(self,
                 hidden_size,
                 src_vocab_size,
                 tar_vocab_size,
                 batch_size,
                 num_layers=1,
                 init_scale=0.1,
                 dropout=None,
                 beam_start_token=1,
                 beam_end_token=2,
                 beam_max_step_num=100)

File: PaddleNLP/PaddleTextGEN/rnn_search/base_model.py (hunk 6 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/base_model.py]: class BaseModel(object):
    def __init__(self,
                 hidden_size,
                 src_vocab_size,
                 tar_vocab_size,
                 batch_size,
                 num_layers=1,
                 init_scale=0.1,
                 dropout=None,
                 beam_start_token=1,
                 beam_end_token=2,
                 beam_max_step_num=100):

Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/base_model.py]: 写入到了__init__()中了，设置为默认值

File: PaddleNLP/PaddleTextGEN/rnn_search/train.py (hunk 9 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/train.py]: 改为 os.path.join()方式

File: PaddleNLP/PaddleTextGEN/rnn_search/reader.py (hunk 16 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/reader.py]: fixed

File: PaddleNLP/PaddleTextGEN/rnn_search/reader.py (hunk 17 above)
Comment (1024er) [reply PaddleNLP/PaddleTextGEN/rnn_search/reader.py]: fixed

**Review by guoshengCS — APPROVED** [review]
LGTM
